syntax = "proto3";

package protocols.navigation;

import "protocols/common/geometry.proto";

message DiscretizedPath {
    repeated PathNode path = 1;
    GoToPoint go_to_point = 2;
}

message PathNode {
    common.Point2Df position = 1;
    common.Point2Df velocity = 2;
    double time = 3;
}

message RotateInPoint {
	common.Point2Df target = 1;
	double target_angle = 2;
	bool clockwise = 3;
	double orbit_radius = 4;
	
	double rotate_velocity = 5;
	double min_velocity = 6;
	double approach_kp = 7;
	double angle_kp = 8;
}

message RotateOnSelf {
	double target_angle = 1;
	common.Point2Df velocity = 2;
	double kp = 3;
}

message GoToPoint {
	common.Point2Df target = 1;
	double target_angle = 2;
	
	MovingProfile moving_profile = 3;
	PrecisionToTarget precision_to_target = 4;
	bool sync_rotate_with_linear_movement = 5;
}

enum PrecisionToTarget {
	High = 1;
	Normal = 2;
}

enum MovingProfile {
	   // Move safely profile
    SafeInStopSpeed = 1;

    // Specific movement profile
    BalancedInApproachBallSpeed = 2;
    BalancedInCarryBallSpeed = 3;
    BalancedInCarryBallLowSpeed = 4;

    // Goto point direct profile
    DirectApproachBallSpeed = 5;
    DirectSafeKickBallSpeed = 6;
    DirectBalancedKickBallSpeed = 7;
    DirectKickBallwithEnemyClose = 8;
    gotoBallWithEnemy = 9;
    PushBallInPenalty = 10;
    TOTOZINHO = 11;

    // Default movement profile
    BalancedInSlowSpeed = 12;
    BalancedInMedianSpeed = 13;
    BalancedInDefaultSpeed = 14;
    BalancedInHighSpeed = 15;

    // Goalkepper profile
    GoalkeeperInTopSpeed = 16;

    // Penalty push ball
    PenaltyPushBall = 17;
}
